File: core/src/main/java/com/gitlab/aecsocket/minecommons/core/Numbers.java
package com.gitlab.aecsocket.minecommons.core;

/**
 * Utilities for numbers.
 */
public final class Numbers {
    private Numbers() {}

    /**
     * Shorthand for squaring a number, equivalent to {@code v * v}.
     * @param v The value.
     * @return The squared value.
     */
    public static double sqr(double v) { return v * v; }

    /**
     * Clamps a value between the minimum and maximum.
     * @param v The value.
     * @param min The minimum.
     * @param max The maximum.
     * @return The clamped value.
     */
    public static int clamp(int v, int min, int max) {
        return Math.min(max, Math.max(min, v));
    }

    /**
     * Clamps a value between the minimum and maximum.
     * @param v The value.
     * @param min The minimum.
     * @param max The maximum.
     * @return The clamped value.
     */
    public static long clamp(long v, long min, long max) {
        return Math.min(max, Math.max(min, v));
    }

    /**
     * Clamps a value between the minimum and maximum.
     * @param v The value.
     * @param min The minimum.
     * @param max The maximum.
     * @return The clamped value.
     */
    public static float clamp(float v, float min, float max) {
        return Math.min(max, Math.max(min, v));
    }

    /**
     * Clamps a value between the minimum and maximum.
     * @param v The value.
     * @param min The minimum.
     * @param max The maximum.
     * @return The clamped value.
     */
    public static double clamp(double v, double min, double max) {
        return Math.min(max, Math.max(min, v));
    }


    /**
     * Clamps a value between 0 and 1.
     * @param v The value.
     * @return The clamped value.
     */
    public static int clamp01(int v) { return clamp(v, 0, 1); }

    /**
     * Clamps a value between 0 and 1.
     * @param v The value.
     * @return The clamped value.
     */
    public static long clamp01(long v) { return clamp(v, 0, 1); }

    /**
     * Clamps a value between 0 and 1.
     * @param v The value.
     * @return The clamped value.
     */
    public static float clamp01(float v) { return clamp(v, 0, 1); }

    /**
     * Clamps a value between 0 and 1.
     * @param v The value.
     * @return The clamped value.
     */
    public static double clamp01(double v) { return clamp(v, 0, 1); }
}

File: core/src/main/kotlin/com/gitlab/aecsocket/alexandria/core/extension/Numbers.kt
package com.gitlab.aecsocket.alexandria.core.extension

import kotlin.math.PI
import kotlin.math.max
import kotlin.math.min

const val EPSILON = 0.000001

fun clamp(value: Double, min: Double, max: Double) = min(max, max(min, value))

fun clamp(value: Float, min: Float, max: Float) = min(max, max(min, value))

fun clamp(value: Int, min: Int, max: Int) = min(max, max(min, value))

fun clamp(value: Long, min: Long, max: Long) = min(max, max(min, value))

fun clamp01(value: Double) = clamp(value, 0.0, 1.0)

fun clamp01(value: Float) = clamp(value, 0f, 1f)

val Int.radians get() = this * (PI / 180)
val Long.radians get() = this * (PI / 180)
val Float.radians get() = this * (PI / 180).toFloat()
val Double.radians get() = this * (PI / 180)

val Int.degrees get() = this * (180 / PI)
val Long.degrees get() = this * (180 / PI)
val Float.degrees get() = this * (180 / PI).toFloat()
val Double.degrees get() = this * (180 / PI)
